event-sourcing/event-source-api/src/main/java/uk/gov/justice/services/eventsourcing/source/api/feed/common/FeedGenerator.java

```diff
@@ -17,48 +23,81 @@
  */
 public class FeedGenerator<E, FE> {
 
-    private final long pageSize;
-
     private final PaginationCapableRepository<E> repository;
 
     private final Entity2FeedEntryMappingStrategy<E, FE> mappingStrategy;
 
-    public FeedGenerator(final long pageSize, final PaginationCapableRepository<E> repository, final Entity2FeedEntryMappingStrategy<E, FE> mappingStrategy) {
-        this.pageSize = pageSize;
+    private final FeedMaxMinProvider<FE> feedMaxMinProvider;
+
+    public FeedGenerator(final PaginationCapableRepository<E> repository,
+                         final Entity2FeedEntryMappingStrategy<E, FE> mappingStrategy,
+                         final FeedMaxMinProvider<FE> feedMaxMinProvider) {
         this.repository = repository;
         this.mappingStrategy = mappingStrategy;
+        this.feedMaxMinProvider = feedMaxMinProvider;
     }
 
-    public Feed<FE> feed(final String page, final UriInfo uriInfo, final Map<String, Object> params) {
-        final long pageNumber = Long.valueOf(page);
-        final List<E> entities = entitiesPage(pageSize, pageNumber, params);
+    public Feed<FE> feed(final long offset,
+                         final Link link,
+                         final long pageSize,
+                         final UriInfo uriInfo,
+                         final Map<String, Object> params) {
+
+        final List<E> entities = entitiesPage(offset, link, pageSize, params);
 
-        return feed(uriInfo, pageNumber, entities);
+        final List<FE> feedEntries = feedEntries(entities, uriInfo);
+
+        return feedWithLinks(link, pageSize, uriInfo, params, feedEntries);
     }
 
-    private Feed<FE> feed(final UriInfo uriInfo, final long pageNumber, final List<E> entities) {
-        final int entityCount = entities.size();
-        return new Feed<>(
-                feedEntries(entities, uriInfo).subList(0, entityCount > pageSize ? entityCount - 1 : entityCount),
-                new Paging(previousPageHrefFrom(uriInfo, pageNumber), nextPageHrefFrom(uriInfo, entities, pageNumber, pageSize)));
+    private Feed<FE> feedWithLinks(final Link link, final long pageSize, final UriInfo uriInfo,
+                                   final Map<String, Object> params, final List<FE> feedEntries) {
+
+        if (feedEntries.isEmpty()) {
+            return new Feed<>(new ArrayList<>(),
+                    new Paging(null, null, pageHref(0, HEAD, pageSize, uriInfo),
+                            pageHref(0, LAST, pageSize, uriInfo)));
+        }
+
+        final long maxSequenceId = feedMaxMinProvider.max(feedEntries);
+        final long minSequenceId = feedMaxMinProvider.min(feedEntries);
+
+        final boolean newEventsAvailable = link != HEAD && recordExists(maxSequenceId + 1, PREVIOUS, pageSize,
+                params);
+
+        final boolean olderEventsAvailable = link != LAST && recordExists(minSequenceId - 1, NEXT, pageSize, params);
+
+        return new Feed<>(feedEntries, new Paging(
```

Comment:
is there anyway we can refactor the paging to be a list/set and keep chaining them? - might make it nicer than having nulls. Just a suggestion, not sure how exactly.

Reply:
Not doing

Reply:
not changing